components/raftstore/src/store/peer.rs

```diff
@@ -1803,11 +1803,14 @@ impl Peer {
             }
         }
         let healthy = self.count_healthy_node(progress.voters());
+        let voters_len = progress.voter_ids().len();
         let quorum_after_change = match ctx.cfg.quorum_algorithm {
             QuorumAlgorithm::IntegrationOnHalfFail => {
-                util::integration_on_half_fail_quorum_fn(progress.voter_ids().len())
+                let majority = raft::majority(voters_len);
+                let custom = util::integration_on_half_fail_quorum_fn(voters_len);
+                cmp::min(voters_len, cmp::max(majority, custom))
```

Comment:
I think this logic should be moved to below because all quorum numbers must be validated.
`quorum_after_change = cmp::min(voters_len, cmp::max(majority_len, quorum_after_change)`

Reply:
Is it really necessary? `majority` is in [majority, voters_len] obviously.

Reply:
It's because the `quorum algorithm` is an enum. But moving it after adding another new quorum function also looks good to me.